### PR TITLE
controller: Reconcile volume to faulted state when all replicas have failed

### PR DESCRIPTION
**Pull request by shuo-wu**
https://github.com/longhorn/longhorn/issues/576

### CODE REVIEW

File: controller/volume_controller.go [hunk 1]
```diff
@@ -11,7 +11,7 @@ import (
 
 	batchv1 "k8s.io/api/batch/v1"
 	batchv1beta1 "k8s.io/api/batch/v1beta1"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (yasker) [inline controller/volume_controller.go]: ?

File: controller/volume_controller.go [hunk 2]
```diff
@@ -383,6 +383,25 @@ func (vc *VolumeController) ReconcileEngineReplicaState(v *longhorn.Volume, e *l
...
 	if e.Status.CurrentState != types.InstanceStateRunning {
+		// Corner case: when vc found healthyCount is 0, vc failed to update volume to faulted state
+		// and volume may be detached (engine stopped state).
+		// Then the volume needs to be reconciled to the faulted state here.
+		failedCount := 0
+		for _, r := range rs {
+			if r.Spec.FailedAt != "" {
+				failedCount++
+			}
+			if failedCount >= v.Spec.NumberOfReplicas {
```
Comment (yasker) [inline controller/volume_controller.go]: It's wrong.

A volume can have 10 failed replicas, but as long as there is one good replica, it's not faulted.

The condition should be no healthy replicas. `FailedAt == "" && HealthyAt != ""`

File: controller/volume_controller.go (hunk 1 above)
Comment (shuo-wu) [reply controller/volume_controller.go]: `go fmt` tool of Goland will automatically add it...

File: controller/volume_controller.go (hunk 2 above)
Comment (shuo-wu) [reply controller/volume_controller.go]: It seems that `HealthyAt != ""` is unnecessary. Here the volume is not running and `HealthyAt` can be empty.

Comment (yasker) [reply controller/volume_controller.go]: Right. The one case may result in `FailedAt == "" && HealthyAt == ""` should be cleaned up by https://github.com/longhorn/longhorn-manager/blob/master/controller/volume_controller.go#L498

**Comment by yasker** [comment]
Please make a comment on the issue about how to test.

**Pull request merged**